mongo/errors.go

```diff
@@ -612,6 +612,56 @@ func (bwe BulkWriteException) HasErrorCodeWithMessage(code int, message string)
 // serverError implements the ServerError interface.
 func (bwe BulkWriteException) serverError() {}
 
+// ClientBulkWriteException is the error type returned by ClientBulkWrite operations.
+type ClientBulkWriteException struct {
+	// A top-level error that occurred when attempting to communicate with the server
+	// or execute the bulk write. This value may not be populated if the exception was
+	// thrown due to errors occurring on individual writes.
+	TopLevelError *WriteError
+
+	// The write concern errors that occurred.
+	WriteConcernErrors []WriteConcernError
+
+	// The write errors that occurred during individual operation execution.
+	// This map will contain at most one entry if the bulk write was ordered.
+	WriteErrors map[int]WriteError
+
+	// The results of any successful operations that were performed before the error
+	// was encountered.
+	PartialResult *ClientBulkWriteResult
+}
+
+// Error implements the error interface.
+func (bwe ClientBulkWriteException) Error() string {
```

Comment:
This function doesn't return an error if the write is unacknowledged. The specifications required that users be able to discern whether a `BulkWriteResult` contains acknowledged results. Either return an error indicating an unacknowledged result, or update `ClientBulkWriteResult` in the spirit of GODRIVER-2821.

Reply:
You can test this with the following:

```go
package main

import (
	"context"

	"go.mongodb.org/mongo-driver/v2/bson"
	"go.mongodb.org/mongo-driver/v2/mongo"
	"go.mongodb.org/mongo-driver/v2/mongo/options"
	"go.mongodb.org/mongo-driver/v2/mongo/writeconcern"
)

func main() {
	client, err := mongo.Connect()
	if err != nil {
		panic(err)
	}

	defer func() { _ = client.Disconnect(context.Background()) }()

	pairs := &mongo.ClientWriteModels{}

	insertOneModel := mongo.NewClientInsertOneModel().SetDocument(bson.D{{"x", 1}})

	opts := options.ClientBulkWrite().SetWriteConcern(writeconcern.Unacknowledged()).SetOrdered(false)

	pairs = pairs.AppendInsertOne("db", "k", insertOneModel)
	_, err = client.BulkWrite(context.Background(), pairs, opts) // Should not panic
	if err != nil {
		panic(err)
	}
}
```

Reply:
Is there a unified spec test that covers this case? If not we should add one / add an integration test.

Reply:
Test case added at https://github.com/mongodb/mongo-go-driver/pull/1884/commits/d10eff2e33a836ef11183ebaa796e5709107ce45#diff-2b04421639d947fd71e799c04cb9df94f58b9be8967998b3e4810c711e9c3671R758-R775